### ISSUE TITLE
Julia v1.5 failure because of @eval in __init__

### ISSUE THREAD

**Issue by blegat** [sisl/CUDD.jl#15]
Loading CUDD in another package gives
```
ERROR: LoadError: LoadError: InitError: Evaluation into the closed module `CUDD` breaks incremental compilation because the side effects will not be permanent. This is likely due to some other module mutating `CUDD` with `eval` during precompilation - don't do this.
Stacktrace:
 [1] eval at ./boot.jl:331 [inlined]
 [2] __init__() at /home/travis/.julia/packages/CUDD/s7EqP/src/ADD_apply.jl:45
```
Because of this line which uses `@eval` in `__init__`:
https://github.com/sisl/CUDD.jl/blob/8bd74c818a1066888f82d9e62dde3f5789049e53/src/ADD_apply.jl#L45
See https://travis-ci.com/github/blegat/Dionysos.jl/jobs/422264066#L607

What's the reason it is used in `__init__`?

**Comment by mykelk** [comment]
@wenting-zhao or @MaximeBouton do you know what's going on here?

**Comment by MaximeBouton** [comment]
hmmm I am not sure to be honest. It seems like it is automatically defining a bunch of functions from the symbol list. 
I don't know if this code was automatically generated or not.

**Comment by mykelk** [comment]
What if we replace line 40 with:
`@eval $(Symbol(name, "_c"))[] = @cfunction($name, Ptr{Node}, (Ptr{Manager}, Ptr{Ptr{Node}}, Ptr{Ptr{Node}}))`

And then remove the `__init__` function?

**Comment by MaximeBouton** [comment]
I think you are right we probably don't need __init__. 

This seems like our use case: https://docs.julialang.org/en/v1/manual/metaprogramming/#Code-Generation

**Comment by guyvdbroeck** [comment]
(I'm also seeing this bug when using CUDD.jl in another package)

**Comment by mykelk** [comment]
My suggestion doesn't actually work. It looks like it works on Julia 1.5 works on travis. It also works on Julia 1.6 on my machine. @guyvdbroeck can you provide more information on your error?

**Comment by guyvdbroeck** [comment]
On Julia 1.6, while checking out https://github.com/Juice-jl/Dice.jl/tree/b2bd16900119386d5e3862a6a1ec21fe9157350f I get the following error if I comment out the workaround `__precompile__(false)`.

```
julia> using Dice
[ Info: Precompiling Dice [94eeadfc-1ba6-40b7-9ed8-dee71557d525]
ERROR: LoadError: LoadError: InitError: Evaluation into the closed module `CUDD` breaks incremental compilation because the side effects will not be permanent. This is likely due to some other module mutating `CUDD` with `eval` during precompilation - don't do this.
Stacktrace:
  [1] eval
    @ ./boot.jl:360 [inlined]
  [2] __init__()
    @ CUDD ~/.julia/packages/CUDD/s7EqP/src/ADD_apply.jl:45
  [3] _include_from_serialized(path::String, depmods::Vector{Any})
    @ Base ./loading.jl:674
  [4] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String)
    @ Base ./loading.jl:760
  [5] _require(pkg::Base.PkgId)
    @ Base ./loading.jl:998
  [6] require(uuidkey::Base.PkgId)
    @ Base ./loading.jl:914
  [7] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:901
  [8] include(mod::Module, _path::String)
    @ Base ./Base.jl:386
  [9] include(x::String)
    @ Dice ~/.julia/dev/Dice/src/Dice.jl:1
 [10] top-level scope
    @ ~/.julia/dev/Dice/src/Dice.jl:9
 [11] include
    @ ./Base.jl:386 [inlined]
 [12] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt64}}, source::Nothing)
    @ Base ./loading.jl:1213
 [13] top-level scope
    @ none:1
 [14] eval
    @ ./boot.jl:360 [inlined]
 [15] eval(x::Expr)
    @ Base.MainInclude ./client.jl:446
 [16] top-level scope
    @ none:1
during initialization of module CUDD
in expression starting at /home/guy/.julia/dev/Dice/src/backend.jl:3
in expression starting at /home/guy/.julia/dev/Dice/src/Dice.jl:1
ERROR: Failed to precompile Dice [94eeadfc-1ba6-40b7-9ed8-dee71557d525] to /home/guy/.julia/compiled/v1.6/Dice/jl_A28Sm2.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] compilecache(pkg::Base.PkgId, path::String, internal_stderr::Base.TTY, internal_stdout::Base.TTY)
   @ Base ./loading.jl:1360
 [3] compilecache(pkg::Base.PkgId, path::String)
   @ Base ./loading.jl:1306
 [4] _require(pkg::Base.PkgId)
   @ Base ./loading.jl:1021
 [5] require(uuidkey::Base.PkgId)
   @ Base ./loading.jl:914
 [6] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:901
```

**Comment by tomerarnon** [comment]
I wasn't aware of this issue, sorry! @mykelk pinged my on it this morning. I wrote the `__init__` function not knowing `@eval` wasn't allowed for precompilation. I think that should be fixed in #17. I tried to precompile `Dice.jl` locally and it seemed to work. Another data point would be useful if someone wants to try as well!

**Comment by guyvdbroeck** [comment]
Precompilation now works for me, much obliged!
